### PR TITLE
Refactor request flow

### PR DESCRIPTION
**Pull request by Namiii**
**TASK**=>[T13471](https://phabricator.omise.co/T13471)

**SUMMARY**
In this PR, we have created a general `Request` that can be passed through the `Client` in order to enable us to have a number of different API calls in the SDK, instead of a hardcoded `TokenRequest`. A base `RequestBuilder` class was also made to be used to create different Requests. Along with this, we have also made changes to the Request callback to return a `Model` instead of a `Token`. In this refactor, a number of classes were also converted to Kotlin.

**QA**
The Token creation process should work as it did before.

**FURTHER WORK**
There are a couple of areas that need to be improved but are segmented off as their own tasks that need to be done after this PR is merged.
- Integrating a Parsing library [T13595](https://phabricator.omise.co/T13595)
- Adding a test flow for Requests [T13596](https://phabricator.omise.co/T13596)

### CODE REVIEW

File: sdk/src/main/java/co/omise/android/api/Client.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            chain.proceed(chain.request()
+                    .newBuilder()
+                    .addHeader("User-Agent", buildUserAgent())
+                    .addHeader("Authorization", Credentials.basic(publicKey, "x"))
+                    .build())
+        }
+    }
+
+    private fun buildUserAgent(): String {
+        return "OmiseAndroid/" + BuildConfig.VERSION_NAME +
```
Comment (pitiphong-p) [inline sdk/src/main/java/co/omise/android/api/Client.kt]: Should we have a settable app name in the User Agent here?

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Client.kt]: Hmmm as in have `OmiseAndroid/` as config too? Think we can but do you have any use case in mind for it? For dev and release maybe?

File: sdk/src/main/java/co/omise/android/api/Invocation.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    private fun processCall(call: TypedCall) {
+        val response = call.execute()
+        if (response.body() == null) {
+            didFail(IOException("HTTP response have no body."))
+            return
+        }
+
+        val rawJson = response.body().string()
+        val model = parseModelFromJson(rawJson, call)
+        if (response.code() in 200..299 && model != null) {
```
Comment (pitiphong-p) [inline sdk/src/main/java/co/omise/android/api/Invocation.kt]: I think we should have a special case for handling 300..399 status codes? We shouldn't allow redirection and have a specific case.

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: So, a custom exception?

Comment (pitiphong-p) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: Correct

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: What would we call this exception btw? :D

File: sdk/src/main/java/co/omise/android/api/RequestBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    /**
+     * Abstract method that needs to be implement by all children of this class to provide response type
+     *
+     * @return Class type of response
+     */
+    protected abstract fun type(): Class<T>
+
+    companion object {
```
Comment (nuxzero) [inline sdk/src/main/java/co/omise/android/api/RequestBuilder.kt]: Shall we use **sealed class** for request method?

File: sdk/src/main/java/co/omise/android/api/Invocation.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        private val replyHandler: Handler,
+        private val httpClient: OkHttpClient,
+        private val request: co.omise.android.api.Request<T>,
+        private val listener: RequestListener<T>
+) {
+
+    fun invoke() {
+        val call = httpClient.newTypedCall(
+                Request.Builder()
+                        .method(request.method,
```
Comment (nuxzero) [inline sdk/src/main/java/co/omise/android/api/Invocation.kt]: Should we make it same line?

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: Which ones?

File: sdk/src/main/java/co/omise/android/api/RequestBuilder.kt (hunk 3 above)
Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/RequestBuilder.kt]: I think we had this discussion for Omise Java as well :) The conclusion was that just having it as String is enough and there is no need to overcomplicate things.

Comment (nuxzero) [reply sdk/src/main/java/co/omise/android/api/RequestBuilder.kt]: Yeah, that mean we need to take this risky. BTW, it's OK.

File: sdk/src/main/java/co/omise/android/api/Invocation.kt (hunk 4 above)
Comment (nuxzero) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: `method()`

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: Hmmm, alright 👌

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: @nuxzero how about this? 

`
        val call = httpClient.newTypedCall(Request.Builder()
                .method(request.method,
                        request.payload)
                .url(request.url)
                .build(),
                request.responseType)
`

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: @nuxzero How about this? 

E5C70AAD-F1E5-401D-B55F-D5075FD15CE5

Comment (nuxzero) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: No no, I mean 
```
val call = httpClient.newTypedCall(
                Request.Builder()
                        .method(request.method, request.payload)
                        .url(request.url)
                        .build(),
                request.responseType)
```

Comment (nuxzero) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: BTW, approved. 👍

Comment (Namiii) [reply sdk/src/main/java/co/omise/android/api/Invocation.kt]: Ahhhh okk, now I get it 👌

**Pull request merged**